broker/src/bench.rs

```diff
@@ -0,0 +1,175 @@
+            let avg_rps = (c_new - skipped) as f64 / dur_secs;
+            (dur, avg_rps)
+        });
+        println!(
+            "sum rps {}, min rps {} total {}, total min {}, duration, avg sum rps {:?}",
+            c_new - c_old,
+            c_min_new - c_min_old,
+            c_new,
+            c_min_new,
+            avg_rps
+        );
+        c_old = c_new;
+        c_min_old = c_min_new;
+    }
+}
+
+fn tli_from_u64(i: u64) -> Vec<u8> {
+    let mut timeline_id = vec![0xFF; 8];
+    timeline_id.extend_from_slice(&i.to_be_bytes());
+    timeline_id
+}
+
+async fn subscribe(client: Option<NeonBrokerClient<Channel>>, counter: Arc<AtomicU64>, i: u64) {
+    let mut client = match client {
+        Some(c) => c,
+        None => NeonBrokerClient::connect("http://[::1]:50051")
+            .await
+            .unwrap(),
+    };
+
+    // let key = SubscriptionKey::All(Empty {});
+    let key = SubscriptionKey::TenantTimelineId(ProtoTenantTimelineId {
+        tenant_id: vec![0xFF; 16],
+        timeline_id: tli_from_u64(i),
+    });
+    let request = SubscribeSafekeeperInfoRequest {
+        subscription_key: Some(key),
+    };
+    let mut stream = client
+        .subscribe_safekeeper_info(request)
+        .await
+        .unwrap()
+        .into_inner();
+
+    while let Some(_feature) = stream.message().await.unwrap() {
+        counter.fetch_add(1, Ordering::Relaxed);
+        // println!("info = {:?}, client {}", _feature, i);
+    }
+}
+
+async fn publish(client: Option<NeonBrokerClient<Channel>>, n_keys: u64) {
+    let mut client = match client {
+        Some(c) => c,
+        None => NeonBrokerClient::connect("http://[::1]:50051")
+            .await
+            .unwrap(),
+    };
+    let mut counter: u64 = 0;
+
+    // create stream producing new values
+    let outbound = async_stream::stream! {
+        loop {
+            let info = SafekeeperTimelineInfo {
+                safekeeper_id: 1,
+                tenant_timeline_id: Some(ProtoTenantTimelineId {
+                    tenant_id: vec![0xFF; 16],
+                    timeline_id: tli_from_u64(counter % n_keys),
+                }),
+                last_log_term: 0,
+                flush_lsn: counter,
+                commit_lsn: 2,
+                backup_lsn: 3,
+                remote_consistent_lsn: 4,
+                peer_horizon_lsn: 5,
+                safekeeper_connstr: "zenith-1-sk-1.local:7676".to_owned(),
+                local_start_lsn: 0,
+            };
+            counter += 1;
+            // println!("sending info = {:?}", info);
+            if counter >= 10 {
+                break;
+            }
+            yield info;
+            time::sleep(Duration::from_millis(1000)).await;
+        }
+    };
+    let response = client.publish_safekeeper_info(Request::new(outbound)).await;
+    println!("pub response is {:?}", response);
+}
+
+#[tokio::main]
```

Comment:
Feels it could be a separate `bin/` module?
Or we could try to move it under `benches/` directory and exclude from the regular compilation cycle entirely (since we do not plan running it regularly)

Reply:
It is already excluded, it is under feature flag which must be explicitly set to get it built. Not sure whether common `benches/` is better...